url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/util/TextRegion.java

```diff
@@ -0,0 +1,239 @@
+            this.length = length;
+            return;
+        } 
+        else if ( ! this.contains( other ) ) {
+            return;
+        }
+        throw new UnsupportedOperationException("Cannot calculate "+this+" MINUS "+other+" , would yield two non-adjactent ranges");
+    }
+    
+    @Override
+    public boolean contains(ITextRegion other)
+    {
+        return other.getStartingOffset() >= this.getStartingOffset() && other.getEndOffset() <= this.getEndOffset();
+    }
+    
+    @Override
+    public boolean overlaps(ITextRegion other)
+    {
+        return this.contains( other ) || other.contains( this ) || 
+           ( ! this.contains( other.getStartingOffset() ) && this.contains( other.getEndOffset() ) ) ||
+           ( this.contains( other.getStartingOffset() ) && ! this.contains( other.getEndOffset() ) );
+    }
+    
+    public void intersect(ITextRegion other) {
+        
+        /*    |-- this --|
+         * |-other-|
+         */
+        if ( ! contains( other.getStartingOffset() ) && contains( other.getEndOffset()-1 ) ) {
+            // ORDER of calculations is important here! 
+            this.length = other.getEndOffset() - this.getStartingOffset();
+            this.startingOffset = other.getStartingOffset();
+            return; 
+        }
+        
+        /*   |-- this --|
+         *     |-other-|
+         */
+        if ( contains( other.getStartingOffset() ) && contains( other.getEndOffset() ) ) {
+            this.startingOffset = other.getStartingOffset();
+            this.length = other.getLength();
+            return; 
+        }
+        
+        /*   |-- this --|
+         *         |-other-|
+         */
+        if ( contains( other.getStartingOffset() ) && ! contains( other.getEndOffset() ) ) {
+            this.length = getEndOffset() - other.getStartingOffset();
+            this.startingOffset = other.getStartingOffset();
+            return; 
+        }   
+        
+        /*   |-- this --|
+         * |-----other-----|
+         */
+        if ( other.contains( getStartingOffset() ) && other.contains( getEndOffset() ) ) {
+            // this range already is the intersection
+            return; 
+        }       
+        throw new IllegalArgumentException( this+" has no intersection with "+other);
+    }
+    
+    @Override
+    public int getEndOffset()
+    {
+        return startingOffset+length;
+    }
+
+    @Override
+    public boolean contains(int offset)
+    {
+        return offset >= this.getStartingOffset() && offset < this.getEndOffset();
+    }
+
+    @Override
+    public boolean isSame(ITextRegion other)
+    {
+        return this == other || ( this.getStartingOffset() == other.getStartingOffset() && this.getLength() == other.getLength() );
+    }
+
+    @Override
+    public String apply(String string)
+    {
+        try {
+            return string.substring( getStartingOffset() , getEndOffset() );
+        } catch(StringIndexOutOfBoundsException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)